README.md

```diff
@@ -0,0 +1,42 @@
+# Introduction
+The 'um-update_toolbox_armhf' is a 'squashfs' minimal root filesystem that can be mounted in early boot stages, with the purpose of providing tools to be able to perform complex setup and update operations. 
+
+This toolbox image is based on Linux Alpine, a security-oriented, lightweight Linux distribution based on musl libc and busybox. 
+
+On top of the standard distro we install the required tooling, such as; e2fsprogs-extra f2fs-tools rsync, resulting in an image with a compressed size of less then 2.6M at time of writing. 
+
+## Repository structure
+Root:
+- License and readme files
+- Docker files used by CI for generating an isolated build environment see...
+- Scripts for building the software
+- Configuration files for Gitlab CI
+
+Scripts folder:
+
+Update or setup scripts, the main entrypoint here is startup.sh. These 
+scripts will be placed in the toolbox '/sbin' folder  
+
+Tests folder:
+- A test to check the build environment
+- A test to validate the generated toolbox functionality
+
+## How to build
+This repository is setup to use an existing docker image from our docker registry at gitlab. The image contains all the tools to be able to build the software. When docker is not installed the build script will try to build on the host machine and will probably fail because the required tools are not installed. If you necessarily wan't to build without docker, read the tests/buildenv.sh script and check what the requirements are. 
+
+By default the build script runs an environment check, builds the image and then validates it by running tests. The first and the latter can be disabled run the help of the build_for_ultimaker.sh script for usage information:  
+```sh
+> ./build_for_ultimaker.sh -h
+    Usage: ./build_for_ultimaker.sh [OPTIONS]
+        -c   Skip run of build environment checks
+        -h   Print usage
+        -t   Skip run of rootfs tests
+```
+
+## Adding update or setup routines to the toolbox
+During the initramfs routine (executed between 2st stage U-boot and 3th stage Linux debian load) this toolbox image will be mounted and a script '/sbin/startup.sh' will be executed. This script should be used to execute whatever required routine. Very important here is to make sure all required resources are nicely cleaned up and the errors are properly handled, not to bring the printer in a unrecoverable state. 
+
+## Updating the docker image in the docker registry
+This process is a bit tricky, we only want to update the docker image when required and we only want to do this after quality checks and when merged to master. 
+
+Docker container changes should be tested locally first. When complete create a commit with the text **dockerfile** in the commit message, this triggers Gitlab CI (when on the master branch) to build and push a new docker container. The new docker container will now be pulled in when executing the build. 
```

Comment:
This should be in confluence to be written more detailed (but also here yes).

Due to missing features of the CI, dockerfile changes cannot be detected and a trigger is used as part of the commit message. When making changes to the docker build environment, tag your message subject with **dockerfile: <subject of the docker change>**.

If for reasons changes are needed on both the environment and the build process, keep the commits unrelated to the environment change (but related to the issue) before the dockerfile change.

This to ensure that the last commit pushed to git contains the trigger word.

If there there are build changes needed AFTER the docker environment change within the same merge request, then this should be noted in the body of the commit message.

For example:
- - -
buildenvironment: fix build due to changes in the build process

Some changes where needed to fix the build process because of a, b and c. These changes rely on previous changes to the dockerfile.
- - -

Note that the latter case should be very rare, as dockerfile changes should be part of the same commit where the build scripts are adjusted accordingly.

One final note, when creating a merge request, and the --logs not used (not common in the web UI), make note that the **dockerfile** was changed as part of the merge request.

While this is cumbersome for now, hopefully this will be fixed by the CI in the future.